### PR TITLE
🎨✅💚: refactor tests for CloudDNS/Zone

### PR DESCRIPTION
**Pull request by LittleFox94**
### Description

Tests are now moved into the package, can be run with mocks instead of the real Engine (which is the default behavior, pass the command line flag "-integration-tests" to the tests to run them against the live Engine API) and, most importantly, generate a random zone name to not conflict with other tests against the Engine in parallel.

Also done:
* `make func-test` now also generates the HTML coverage report as `make test` already does 
* extracted `pkg/utils/test.RandomHostname()` from `vsphere` tests to also use in `clouddns` tests

### Release Note
Release note for [CHANGELOG](https://github.com/anexia-it/go-anxcloud/blob/master/CHANGELOG.md):

```release-note
NONE
```

### References

References all the failing integration tests.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

**Comment by toothstone** [comment]
Hi, I can't seem to get the CloudDNS tests to run in integration test mode, `make func-tests` actually has them passing even if there is no ANEXIA_TOKEN set. Am I holding it wrong? I think it would be helpful to put the exact commands into the README here, we can't expect people to read the Makefile to find out about some feature flag.

Also the phrasing of the PR and commit description is somewhat misleading, to me it reads as if the `--integration-tests` flag is activating the mocks.

Lastly, as a general note, there is quite a lot of stuff happening within this one PR, which makes it hard to review. Going forward it would be nice to have separate feature PRs (new mocks) and refactoring PRs (move to pkg, introduce randomness, generalize hostname generator) where possible, or at least some more dedicated commits.

**Comment by LittleFox94** [comment]
> `make func-tests` actually has them passing even if there is no ANEXIA_TOKEN set

hm this is expected to fire against the real API, will look into it - thanks o.o

> Also the phrasing of the PR and commit description is somewhat misleading, to me it reads as if the `--integration-tests` flag is activating the mocks.

Will change

> Lastly, as a general note, there is quite a lot of stuff happening within this one PR, which makes it hard to review. Going forward it would be nice to have separate feature PRs (new mocks) and refactoring PRs (move to pkg, introduce randomness, generalize hostname generator) where possible, or at least some more dedicated commits.

Hm true .. sorry, will take more care in the future

**Comment by LittleFox94** [comment]
There was indeed a bug in running the CloudDNS against the real Engine - this is fixed now and fails as intended without `ANEXIA_TOKEN` set in the environment.

**Comment by LittleFox94** [comment]
The coverage decrease is due to updating a newly creating zone failing with a server error (ENGSUP-5233):
https://github.com/anexia-it/go-anxcloud/pull/64/files#diff-ee83c858f2de755dae95084e93f0416bd100c0180c8b1a85a92dfdf903150d7aR223

File: Makefile [hunk 1]
```diff
@@ -33,7 +33,16 @@ test:
 
 .PHONY: func-test
 func-test:
-	CGO_ENABLED=1 go test -coverpkg ./pkg/... -coverprofile coverage.out -timeout 180m ./tests/...
+	# we run the tests in our tests directory and for every package listed additionally
+	# some of the newer tests (starting with the refactored CloudDNS tests) can run as
+	# either unit or integration tests, using mocks of the API for unit tests. They need
+	# the --integration-tests flag to run as integration tests.
```
Comment (toothstone) [inline Makefile:39]: flag looks different below

**Review by toothstone — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (LittleFox94) [reply Makefile:39]: I wanted to point out that go does not differentiate between `--integration-tests` and `-integration-tests`, but the flag is not actually plural - good catch x)

**Review by toothstone — APPROVED** [review]
LGTM

We need a general overhaul of our testing instruction, I've created https://ats.anexia-it.com/browse/SYSENG-682 to track that. To keep things moving, I'm fine with merging this PR as it is now.

**Comment by LittleFox94** [comment]
Thanks for writing that issue :)

Merging now

**Pull request merged**